feat: Use local Headscale instance for molecule testing

mprasil opened the pull request:
This adds Headscale container setup to Molecule scenarios to be used for testing when `TAILSCALE_CI_KEY` is not set to enable easier local testing.

If the env variable is set, Molecule tests will revert back to the previous behavior and use Tailscale key for testing.

artis3n commented:
Thanks for this PR! I want to move some things around, but I like this idea and I'd like to explore it.

Please take a look at the CI failures (you should be able to reproduce by running locally with the Makefile commands) and fix.

I wrote the molecule tests intending them to be acceptance tests always talking to Tailscale, but I like the idea of supporting Headscale compatibility in the role (I've not used Headscale myself) and therefore in the tests as well.

I'd prefer a redesign following this format:
- Tests, and the role, talk to Tailscale by default.
- Tests fail if `TAILSCALE_CI_KEY` is not set.
- Tests can be run with some env var that changes the default behavior to point to Headscale instead. This should be a different env var, otherwise the tests might silently pass against Headscale when we intend them to run against Tailscale. If this env var is present, we don't need to check for the presence of `TAILSCALE_CI_KEY`.
- If the Headscale docker container and config are the same for every test, I'd prefer it to exist in 1 location in the repo and be referenced by each test, instead of copying the same files throughout the molecule directories.
- The ansible role successfully starts a server through Headscale instead of tailscale (is that a flag on tailscale up or something? I'm not familiar with how to use Headscale)

Let me know if you are interested in working on this structure for this PR. Otherwise, we can move it to an issue and I will take some time to dig into it, but it won't be for a little while.

mprasil commented:
Thanks for feedback. I'm definitely open to continue working on this. I don't use Tailscale myself so instead of creating an account for testing I took the "lazy" approach of adding support for Headscale. 😄

> If the Headscale docker container and config are the same for every test, I'd prefer it to exist in 1 location in the repo and be referenced by each test, instead of copying the same files throughout the molecule directories.

I'll investigate this further, but upon cursory look I didn't see obvious way to do that with Molecule. (I remember it was possible with older versions, so perhaps I'll find a way, I don't like the repetition either)

> The ansible role successfully starts a server through Headscale instead of tailscale (is that a flag on tailscale up or something? I'm not familiar with how to use Headscale)

Yes, it's a flag. You just need to pass URL for Headscale with `--login-server=`

It should not be much work to drive this via an extra en variable. Your concern with role silently passing against local Headscale makes sense to me.

mprasil commented:
I have now added [2d1c033](https://github.com/artis3n/ansible-role-tailscale/pull/319/commits/2d1c033225e6278475b51416b7794b0b09740903) that modifies the logic so that Headscale is never used automatically and must be explicitly enabled with `USE_HEADSCALE`.

As for the CI failures, unfortunately I can't reproduce this locally and CI does not show why it fails due to `no_log`. The failing step seems to be waiting for container to start up. @artis3n do you have any more access to investigate what it could be?

artis3n commented:
Hm, I don't have any further details as to the CI failure - it looks like Molecule is failing setting up the Headscale container. I will try to look into this in further detail in a few days. Does running `make test-default` locally work for you? That is the same test that is running here. The make test commands expect `TAILSCALE_CI_KEY but you can comment out the line or set the env var to an empty string.

https://github.com/artis3n/ansible-role-tailscale/blob/9a19bf1e94c76149af7b1700f9060c6fe677eda9/Makefile#L40

artis3n commented:
Under Files Changed I do see some annotated `ansible-lint` errors across a few of the files, so those errors may be influencing Molecule breaking down - if ansible-lint is interpreting these files incorrectly, molecule probably is too.

mprasil commented:
I have reworked slightly how Headscale is spun up. Now instead of building custom image, it uses upstream image directly and mounts config file inside the container instead. This way I could avoid having `Dockerfile` completely and the config file now can be shared across scenarios. So no more duplicate files. 🎉 

I have also addressed the lint issues in [72d959b](https://github.com/artis3n/ansible-role-tailscale/pull/319/commits/72d959b86dd9e9a6d590d65050b7c0a49c17f5df) - ansible-lint assumed the headscale config yaml is actually playbook and unsurprisingly reported failure trying to parse it.

Hopefully this fixes the CI issues, I have no idea what else could cause this.

> Does running make test-default locally work for you? 

Yes, I ran `TAILSCALE_CI_KEY=fake USE_HEADSCALE=true make test-default` fine with the current code.

Let's try to re-run it again now that we use upstream image and see if it still fails.

artis3n commented:
Great, tests are passing now! For the `ansible-lint` step, it seems you should add `molecule/default/init_tailscale_vars.yml` to the molecule exclude list too

artis3n commented on molecule/default/molecule.yml:
```diff
@@ -11,9 +11,22 @@ platforms:
+  - name: headscale
+    image: ${HEADSCALE_IMAGE:-headscale/headscale:latest}
+    command: headscale serve
+    pre_build_image: true
```
Do you still use `pre_build_image: true` on images that aren't pre-configured to run Molecule? 🤔

mprasil replied:
Yes. This just tells Molecule not to build docker image. (which is the default behavior) There is no specific requirement for the images to be "built for molecule". Molecule itself does not really do anything with the containers other than starting them, it's up to the later prepare/converge/verify/.. playbooks to do something with the hosts if needed. But molecule itself does not care about that at all.

artis3n commented on molecule/args/molecule.yml:
```diff
@@ -11,11 +11,26 @@ platforms:
+    networks:
+      - name: headscale
+    volumes:
+      - "${MOLECULE_PROJECT_DIRECTORY}/molecule/default/headscale.config.yaml:/etc/headscale/config.yaml"
```
Nice! While this might add some confusion for the non-default scenarios (test artifacts referenced outside the scope of the scenario) I prefer this for maintainability of the headscale config.

mprasil replied:
Yeah, I've seen the pattern out there where people [refer common files from default scenario](https://github.com/search?q=path%3Amolecule%2F*%2Fmolecule.yml+NOT+path%3Adefault+%22molecule%2Fdefault%22&type=code) to avoid code repetition, so hopefully it's not too confusing.

artis3n commented on molecule/default/headscale.config.yaml:
```diff
@@ -0,0 +1,32 @@
+# minimal Headscale configuration for local testing
```
Since I've never used Headscale, mind including a link to Headscale's config docs at the top of here too?

artis3n commented on molecule/default/molecule.yml:
```diff
@@ -11,9 +11,22 @@ platforms:
     command: ${MOLECULE_COMMAND:-/usr/sbin/init}
     volumes:
       - /sys/fs/cgroup:/sys/fs/cgroup:rw
+    docker_networks:
```
Up in the `dependency:` section, `community.general` is included in our `requirements-file` but `community.docker`, used only for the `prepare.yml` molecule headscale tasks, is not currently in `requirements.yml`. I believe `community.docker` should be added to `requirements.yml`, as that should only impact local developers testing and working on the role, not consumers of the role. <https://github.com/artis3n/ansible-role-tailscale/blob/main/meta/main.yml#L53> is what the production role cares about. (taking a note for myself to confirm before merging)

Please follow the pattern with `community.general` and use the Git URL reference for `community.docker`, which bypasses timeout issues Ansible Galaxy regularly experiences - https://github.com/artis3n/ansible-role-tailscale/pull/288

mprasil replied:
Yes, you're right, I'm going to add it there. It ended up working fine, because molecule itself depends on `community.docker`, but it's definitely good idea to explicitly state the dependency in requirements file.

artis3n commented on molecule/vars-priority/converge.yml:
```diff
@@ -1,10 +1,11 @@
-  hosts: all
+  hosts: instance
   vars:
-    tailscale_authkey: "{{ lookup('env', 'TAILSCALE_CI_KEY') }}"
```
So this test is intentionally testing `vars` vs. `defaults` and that vars defined in the playbook scope here have correct precedence and are correctly picked up by the role (there was an issue with that way back when). I want to maintain that test - perhaps we add `tailscale_args ` here instead?

mprasil replied:
What are you _actually_ trying to test here? The variable precedence is something that [Ansible itself guarantees](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_variables.html#understanding-variable-precedence) and hopefully tests before releasing new version. If this test stopped working, it would be bug in Ansible. Or am I missing something?

artis3n reviewed (commented):
Thanks for your work on this PR! This looks useful so far.

Can you add a `headscale` Molecule directory/scenario that always runs against Headscale? Unlike the other molecule scenarios where you've added behavior that defaults to run against Tailscale but can run against Headscale instead, I'd like a reproduction of the `default` scenario that just always runs against Headscale so, alongside the regular testing against Tailscale, there's a test ensuring Headscale compatibility running every time in CI.

Then we can add that new scenario as a new job in <https://github.com/artis3n/ansible-role-tailscale/blob/main/.github/workflows/pull_request_target.yml>.

mprasil commented:
Added few more commits to address issues raised in review. Also added another CI job to test against Headscale. I'm not super familiar with GitHub workflows configuration, but I think that should be correct. Please double-check.

artis3n commented on .github/workflows/molecule.yml:
```diff
@@ -59,3 +59,41 @@ jobs:
           MOLECULE_COMMAND: "${{ inputs.command }}"
           TAILSCALE_CI_KEY: "${{ secrets.tailscale_key }}"
+
+  run-molecule-headscale:
```
Sorry, I should have been clearer - this looks good, but please put it at the bottom of the `pull_request_target.yaml` file:

```yaml
  run-molecule-headscale:
    environment: E2E
    runs-on: ubuntu-22.04

    steps:
      - uses: actions/checkout@v3
        with:
          ref: ${{ github.event.pull_request.head.sha }}

      - name: Install dependency manager
        run: pipx install poetry

      - name: Set up Python 3.x
        id: setup-python
        uses: actions/setup-python@v4.5.0
        with:
          python-version: "3.11"
          cache: 'poetry'

      - name: Install packages
        run: poetry install --no-interaction

      - name: Default scenario with Headscale
        run: poetry run molecule test --scenario-name default
        env:
          TAILSCALE_CI_KEY: "unused"
          USE_HEADSCALE: "true"

      - name: Uninstall scenario with Headscale
        run: poetry run molecule test --scenario-name state-absent
        env:
          TAILSCALE_CI_KEY: "unused"
          USE_HEADSCALE: "true"
```

mprasil replied:
Moved. You're right, it does not make sense to test all operating systems against Headscale as well as there should be no practical difference in how this is configured between different OS versions.

artis3n reviewed (commented):
Just move the test around and then we will merge! Thanks for being flexible and responsive in this PR!

mprasil commented on .github/workflows/pull_request_target.yml:
```diff
@@ -176,3 +176,37 @@ jobs:
+
+  run-molecule-headscale:
+    environment: E2E
+    runs-on: ${{ inputs.runner }}
```
Ah, missed that one. 😅

artis3n commented:
Yes! Will be publishing a release within the next few days with these changes. Thanks for your contributions 🙂
